Implementation of SuperPoint and AutoModelForKeypointDetection

sbucaille opened the pull request:
# What does this PR do?

This PR implements SuperPoint, one of the few models that generate keypoints and descriptors given an image, as discussed in [this previous pull request](https://github.com/huggingface/transformers/pull/25697)
The goal is to implement this model and a new type of AutoModel : `AutoModelForKeypointDetection` (name to be discussed).
This PR is also the replacement of a previous [PR](https://github.com/huggingface/transformers/pull/25786) for which the branch ended up being unusable.

## Who can review?

@amyeroberts @ArthurZucker @rafaelpadilla 

## TODO's

- [x] Implement SuperPointConfig and SuperPointModel as PretrainedConfig and PretrainedModel
- [x] Generate a conversion script for the original weights
- [x] Implement the new `AutoModelForKeypointDetection` mapping
- [x] Test the model
- [x] Write documentation

Commit:
Commit to create and fix add_superpoint branch with a new branch

sbucaille commented:
Hi @amyeroberts @rafaelpadilla @ArthurZucker ,

I couldn't see other solution than to create a branch from scratch and re-implementing everything that was discussed since the beginning of the original [PR](https://github.com/huggingface/transformers/pull/25786). Here, most of my problems (RegNet related error when using `make fix-copies` which was caused by the fact that my environment was outdated after doing the merge, the `index.md` huge diff, and others...) are fixed.
As I said, I re-implemented everything that was in the other PR, and addressed all other points that were mentioned in reviews (such as docs) which of course may need more discussion. I tried to make sure everything point discussed were included.
Anyway, I wanted to implement this model in transformers and contribute, you told me it would take more time than the other method, I'll take all this mess I made myself as part of the learning path 😆
Let me know what you think, what should be the next steps, I believe we're almost there ! 🤗 

Steven

Commit:
Added SuperPoint docs

amyeroberts commented:
Hi @sbucaille, thanks for opening this PR and working from the previous reviews! 

At the moment, I don't see any tests implemented for the model or image processor. The next step would be adding those 🤗

Commit:
Added tests

Commit:
Removed commented part

sbucaille commented:
Hi @amyeroberts @ydshieh ,
I just looked again at the PR and noticed that there were conflicts due to recent changes to the `main`branch. So instead of making the merge directly through GitHub interface, I did what you advised me to do earlier, I updated my main branch and rebased this branch onto the main, and managed to resolve the conflicts (the implementation of StableLM ended up at the same lines SuperPoint is implemented in all the readme's and __init__'s), then I pushed the changes.
First question, is it normal that this PR lists all the other commits ?
Second question, if I merged this way, why is GitHub still complaining and asking me to merge manually through the interface like such : 
![image](https://github.com/huggingface/transformers/assets/24275548/7b3306f2-b34f-43a8-bd8b-c2a5c8aa8902)
In the end I feel like merging manually through the interface would have been faster. Am I missing something ?

Also, appears all the tests pass now 🤗

amyeroberts commented:
Hi @sbucaille, regarding your questions about rebasing

> First question, is it normal that this PR lists all the other commits ?

No, it shouldn't. Did you force push after rebasing? It's necessary to do this, as rebasing is effectively rewriting the history of  the branch, 

> Second question, if I merged this way, why is GitHub still complaining and asking me to merge manually through the interface like such :

What do you mean by "the interface"? 

When performing a rebase, it might be necessary to resolve conflicts. What rebasing is trying to do is move the first commit of your branch to the head of main. However, as the branch was originally branched off an older commit, there might be conflicts if files in this branch make changes to files which have also been modified on `main`. It might also happen if you've also merged in `main` into this branch in the interim, as the branch now contains information about newer, upstream commits on `main`. Therefore, how to rewrite the history, so this branch and all its commits are from the tip of `main` isn't obvious. 

The easiest way to manage this is to rebase main into the branch often, have short-lived branches and avoid merge commits in the branch.

Here are some nice articles about rebasing which I found useful when first learning about it: 
* https://www.atlassian.com/git/tutorials/merging-vs-rebasing
* https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase

sbucaille commented:
@amyeroberts Alright I think I start to grasp the base and merge things, this PR is my first in any open source project and I knew in advance I would have been in front of such issues, so thank you for taking the time 🤗
As of now, is my PR still ok ? Should we continue to review the changes ? What are the next steps ?

amyeroberts commented:
@sbucaille I'd suggest first resolving the git history, so this branch only contains commits that are part of this PR. Once our house is tidy we can review and iterate on the PR. From first glance, things are looking good! Doing this now will be a lot easier than trying to do it in the middle of a review.

sbucaille commented:
@amyeroberts holy that was not easy but I think I understood the mechanism, I once more rebased the branch to include the latest commits. It seems I am now 4 commits ahead of main, which should be fine (until tomorrow morning where new commits will be pushed to main I guess).
So I guess it is better now to see a bunch of "force push" in the pull request than the list of all the commits from the main right ? 
How often do you suggest to do this manipulation ? You talked about short lived branches but I guess this does not apply to mine since I've been working on it for a couple of months right ?
Anyway, I'm ready to move on the last review process

Commit:
Fixed dummy_pt_objects

amyeroberts commented:
@sbucaille Great! So main will continue to have lots of new commits. The history you're seeing isn't so much number of commits "ahead" of main, rather, the number of commits on top of a certain main commit, let's say ABC. When you rebase, you're re-writing so that the commits are now on top of the latest main commit, say EDF. It's kind of like shifting things along in history. 

You don't need to rebase every time there's new commits on main. The purpose of rebasing is to make sure your branch and main don't diverge and any changes made to the same file has no clashes and a clear order to apply commits. As well as including any important upstream changes e.g. bug fixes. It's basically the same as having a merge commit but a lot cleaner as only the branches commits are shown.

> So I guess it is better now to see a bunch of "force push" in the pull request than the list of all the commits from the main right ?

Yep! This just lets me know there's been a rewrite of the history. This is important as rebasing can also be used for editing the commits in the branch itself.

> How often do you suggest to do this manipulation ? You talked about short lived branches but I guess this does not apply to mine since I've been working on it for a couple of months right ?

It depends. If you touch a lot of common files, you'll want to do it regularly. Particularly in the final steps when getting ready to merge the branch in - at least once a day. This is a model PR, so probably doesn't need it quite so often. Now you've done the difficult rebase, all the rest should be pretty easy to do. In fact, I'd image there would be little to no conflicts. 

We'd like all branches to live as short as possible. This is less true for model PRs - but we still would like them to be resolved in the order of days/weeks wherever possible. 

> Anyway, I'm ready to move on the last review process

Great! Next step is to get all the tests passing on the CI.

Commit:
Committed missing files

Commit:
Fixed README.md

sbucaille commented:
@amyeroberts Seems right, now. Turns out I uncommitted certain files during my rebase, I'll need to practice a bit more in the future 😅 Tests are passing !

amyeroberts commented on README_fr.md:
```diff
@@ -485,6 +485,7 @@ Nombre actuel de points de contrôle : ![](https://img.shields.io/endpoint?url=h
 1. **[Splinter](https://huggingface.co/docs/transformers/model_doc/splinter)** (de l'Université de Tel Aviv), publié dans l'article [Réponse à quelques questions avec peu d'exemples par la pré-sélection des spans](https://arxiv.org/abs/2101.00438) par Ori Ram, Yuval Kirstain, Jonathan Berant, Amir Globerson, Omer Levy.
 1. **[SqueezeBERT](https://huggingface.co/docs/transformers/model_doc/squeezebert)** (de Berkeley) a été publié dans l'article [SqueezeBERT : Que l'apprentissage automatique peut-il apprendre au traitement du langage naturel sur les réseaux neuronaux efficaces ?](https://arxiv.org/abs/2006.11316) par Forrest N. Iandola, Albert E. Shaw, Ravi Krishna et Kurt W. Keutzer.
 1. **[StableLm](https://huggingface.co/docs/transformers/model_doc/stablelm)** (from Stability AI) released with the paper [StableLM 3B 4E1T (Technical Report)](https://stability.wandb.io/stability-llm/stable-lm/reports/StableLM-3B-4E1T--VmlldzoyMjU4?accessToken=u3zujipenkx5g7rtcj9qojjgxpconyjktjkli2po09nffrffdhhchq045vp0wyfo) by  Jonathan Tow, Marco Bellagente, Dakota Mahan, Carlos Riquelme Ruiz ...
```
There's some funny deletions happening here

sbucaille replied:
I think it is fixed with my latest rebase

sbucaille replied:
Ok I've been looking a long time at this file now and I can't see the difference between my version and the main one 😅 . Is it because of the red color over the text ? 
![image](https://github.com/huggingface/transformers/assets/24275548/7bd8799b-a99e-4e18-8bf9-b0034b3494b9)
If so, then the same marks are visible in the main [README file](https://github.com/huggingface/transformers/blob/main/README_fr.md?plain=1#L490) 
I've been looking at the diff in PyCharm which I'm using and I can't see any difference apart from my SuperPoint line : 
![image](https://github.com/huggingface/transformers/assets/24275548/9378981d-168f-4a6c-9df3-3d40a9e467c2)

Please guide me on this 😅

amyeroberts replied:
Yes, it's because of the red colour o the text. It's possible there's something funny being triggered on github's side, but I'd expect to see it in other PRs. 

I suspect there's something funny happening with the text encoding. What I would suggest is reverting this file back to the one one main, and then running make fix-copies again to have the README updated manually. I'd try to commit the changes without actively opening and saving the file, as the diff issues might be creeping in on this step. 

```
git checkout -- README_fr.md
make fix-copies
```

amyeroberts commented on README_zh-hant.md:
```diff
@@ -448,7 +448,8 @@ conda install conda-forge::transformers
 1. **[Splinter](https://huggingface.co/docs/transformers/model_doc/splinter)** (from Tel Aviv University) released with the paper [Few-Shot Question Answering by Pretraining Span Selection](https://arxiv.org/abs/2101.00438) by Ori Ram, Yuval Kirstain, Jonathan Berant, Amir Globerson, Omer Levy.
 1. **[SqueezeBERT](https://huggingface.co/docs/transformers/model_doc/squeezebert)** (from Berkeley) released with the paper [SqueezeBERT: What can computer vision teach NLP about efficient neural networks?](https://arxiv.org/abs/2006.11316) by Forrest N. Iandola, Albert E. Shaw, Ravi Krishna, and Kurt W. Keutzer.
-1. **[StableLm](https://huggingface.co/docs/transformers/model_doc/stablelm)**  released with the paper [StableLM 3B 4E1T (Technical Report)](https://stability.wandb.io/stability-llm/stable-lm/reports/StableLM-3B-4E1T--VmlldzoyMjU4?accessToken=u3zujipenkx5g7rtcj9qojjgxpconyjktjkli2po09nffrffdhhchq045vp0wyfo) by  Jonathan Tow, Marco Bellagente, Dakota Mahan, Carlos Riquelme Ruiz, Duy Phung, Maksym ...
+1. **[StableLm](https://huggingface.co/docs/transformers/model_doc/stablelm)** (from Stability AI) released with the paper [StableLM 3B 4E1T (Technical Report) by Jonathan Tow, Marco Bellagente, Dakota Mahan, Carlos Riquelme Ruiz, Duy Phung, Maksym Zhuravinskyi, Nathan Cooper, Nikhil Pinnaparaju, Reshinth Adithyan, and James Baicoianu.
```
This PR shouldn't change the StableLM entry

sbucaille replied:
This should also be fixed with the latest rebase

amyeroberts commented on docs/source/en/model_doc/superpoint.md:
```diff
@@ -0,0 +1,127 @@
+<!--Copyright 2023 The HuggingFace Team. All rights reserved.
```
nit

```suggestion
<!--Copyright 2024 The HuggingFace Team. All rights reserved.
```

amyeroberts commented on src/transformers/modeling_outputs.py:
```diff
@@ -1751,3 +1751,38 @@ def logits(self):
         return self.reconstruction
+
+
+@dataclass
```
Let's move this to the modeling file. At the moment we don't have other models using and it's highly specific to this one model

sbucaille replied:
Fixed in [this commit](https://github.com/huggingface/transformers/pull/28966/commits/0d2c3a11bc9c36557546633a69f6004c0948798e)

amyeroberts commented on src/transformers/models/superpoint/image_processing_superpoint.py:
```diff
@@ -0,0 +1,279 @@
+            image (`np.ndarray`):
+                Image to resize.
+            size (`Dict[str, int]`):
+                Dictionary of the form `{"shortest_edge": int}`, specifying the size of the output image. If
+                `size["shortest_edge"]` >= 384 image is resized to `(size["shortest_edge"], size["shortest_edge"])`.
+                Otherwise, the smaller edge of the image will be matched to `int(size["shortest_edge"] / crop_pct)`,
+                after which the image is cropped to `(size["shortest_edge"], size["shortest_edge"])`.
```
This doesn't match the default size format with `"height"` and `"width"`

sbucaille replied:
Fixed in [this commit](https://github.com/huggingface/transformers/pull/28966/commits/a5b05900dfee957d88db3e28072e9edace7c68ef)

amyeroberts commented on src/transformers/models/superpoint/image_processing_superpoint.py:
```diff
@@ -0,0 +1,279 @@
+        """
+        size = get_size_dict(size, default_to_square=False)
+
+        resize_size = get_resize_output_image_size(
+            image, size=(size["height"], size["width"]), input_data_format=input_data_format
+        )
```
This isn't needed if we have "height" and "width" in the dictionary

sbucaille replied:
Fixed in [this commit](https://github.com/huggingface/transformers/pull/28966/commits/a5b05900dfee957d88db3e28072e9edace7c68ef)

amyeroberts commented on src/transformers/models/superpoint/modeling_superpoint.py:
```diff
@@ -0,0 +1,469 @@
+        return descriptors
+
+    @staticmethod
+    def __sample_descriptors(keypoints, descriptors, scale: int = 8) -> torch.Tensor:
```
nit

```suggestion
    def _sample_descriptors(keypoints, descriptors, scale: int = 8) -> torch.Tensor:
```

amyeroberts commented on src/transformers/models/superpoint/image_processing_superpoint.py:
```diff
@@ -0,0 +1,279 @@
+            input_data_format = infer_channel_dimension_format(images[0])
+
+        # Checking if image is RGB or grayscale
+        for i in range(len(images)):
+            if not is_grayscale(images[i], input_data_format):
+                images[i] = convert_to_grayscale(images[i], input_data_format=input_data_format)
```
Given this logic - you should add tests to make sure both RGB images are correctly converted to grayscale. I believe the image processor won't currently work with grayscale inputs (this is a global issue in the library)

sbucaille replied:
Actually not sure on this one what exactly you expected, I added tests to ensure the image processor properly converts to grayscale, by checking the values of the 3 channels, let me know if it is correct. (in [this commit](https://github.com/huggingface/transformers/pull/28966/commits/8c1bb22438740ba37e99bd77f94093b9c71f8cf4))

amyeroberts commented on docs/source/en/model_doc/superpoint.md:
```diff
@@ -0,0 +1,127 @@
+
+The outputs contain the list of keypoint coordinates with their respective score and description (a 256-long vector).
+
+You can also feed multiple images to the model. Due to the nature of SuperPoint to output a dynamic number of keypoint,
```
```suggestion
You can also feed multiple images to the model. Due to the nature of SuperPoint, to output a dynamic number of keypoints,
```

amyeroberts commented on docs/source/en/model_doc/superpoint.md:
````diff
@@ -0,0 +1,127 @@
+You can also feed multiple images to the model. Due to the nature of SuperPoint to output a dynamic number of keypoint,
+you will need to use the mask attribute to retrieve the respective information :
+
+```python
````
Nice example :)

sbucaille replied:
Thanks ! Keypoints are better understood when they overlay the image, I thought it would be worth it to add some visualization in the example :)

amyeroberts commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -1207,6 +1208,12 @@
     ]
 )
 
+MODEL_FOR_KEYPOINT_DETECTION_MAPPING_NAMES = OrderedDict(
+    [
+        ("superpoint", "SuperPointModel"),
+    ]
+)
```
I realise this was originally my suggestion, but reflecting and looking at this again I think we don't need to add this special class. We can remove `MODEL_FOR_KEYPOINT_DETECTION_MAPPING_NAMES` and just use `AutoModel` where needed

sbucaille replied:
I believe this is fixed with [this commit](https://github.com/huggingface/transformers/pull/28966/commits/4c3cc2e92f1edc5f0c5b7f4db679b328f56aac48)

amyeroberts commented on src/transformers/models/superpoint/modeling_superpoint.py:
```diff
@@ -0,0 +1,469 @@
+
+        return keypoints, scores
+
+    def __get_pixel_scores(self, encoded: torch.Tensor) -> torch.Tensor:
+        """Based on the encoder output, compute the scores for each pixel of the image"""
+        scores = self.relu(self.conv_score_a(encoded))
+        scores = self.conv_score_b(scores)
+        scores = nn.functional.softmax(scores, 1)[:, :-1]
+        batch_size, _, height, width = scores.shape
+        scores = scores.permute(0, 2, 3, 1).reshape(batch_size, height, width, 8, 8)
+        scores = scores.permute(0, 1, 3, 2, 4).reshape(batch_size, height * 8, width * 8)
+        scores = simple_nms(scores, self.nms_radius)
+        return scores
+
+    def __extract_keypoints(self, scores: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```
nit

```suggestion
    def _get_pixel_scores(self, encoded: torch.Tensor) -> torch.Tensor:
        """Based on the encoder output, compute the scores for each pixel of the image"""
        scores = self.relu(self.conv_score_a(encoded))
        scores = self.conv_score_b(scores)
        scores = nn.functional.softmax(scores, 1)[:, :-1]
        batch_size, _, height, width = scores.shape
        scores = scores.permute(0, 2, 3, 1).reshape(batch_size, height, width, 8, 8)
        scores = scores.permute(0, 1, 3, 2, 4).reshape(batch_size, height * 8, width * 8)
        scores = simple_nms(scores, self.nms_radius)
        return scores

    def _extract_keypoints(self, scores: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```

amyeroberts commented on src/transformers/models/superpoint/modeling_superpoint.py:
```diff
@@ -0,0 +1,469 @@
+
+
+class SuperPointConvBlock(nn.Module):
+    def __init__(self, in_channels: int, out_channels: int, add_pooling: bool = False) -> None:
```
All custom layers should accept the config as their first argument

sbucaille replied:
I am not sure what to do here, because if I add config as argument, it will not be used by the module because it relies only on `in_channels`, `out_channels` and `add_pooling`. But it can't rely exclusively on config since the SuperPointConvBlock can't know what values to pick from `encoder_hidden_sizes` ? 🤔

amyeroberts replied:
That's OK, it should still have it as its first argument though

sbucaille replied:
Fixed in [this commit](https://github.com/huggingface/transformers/pull/28966/commits/41ee72af5e79860da85db4089c1f0174ac3c7a42)

amyeroberts commented on src/transformers/models/superpoint/modeling_superpoint.py:
```diff
@@ -0,0 +1,469 @@
+        config_class=_CONFIG_FOR_DOC,
+        modality="vision",
+    )
+    def forward(
```
By convention, the model should still accept `labels` as an input. We can raise an exception immediately in the forward pass if it's no `None` saying training isn't possible with this model

sbucaille replied:
Added this in [this commit](https://github.com/huggingface/transformers/pull/28966/commits/d36c58e0b9ae77703e876b5bbf72ec98df90c444). I added tests to ensure we raise the error when labels are indeed not None

amyeroberts commented on tests/models/superpoint/test_modeling_superpoint.py:
```diff
@@ -0,0 +1,302 @@
+    @slow
+    def test_inference(self):
+        model = SuperPointModel.from_pretrained("stevenbucaille/superpoint").to(torch_device)
+        self.infer_on_model(model)
```
No need to add a separate method here - just put all the logic from `infer_on_model` here

sbucaille replied:
Changed it in [this commit](https://github.com/huggingface/transformers/pull/28966/commits/d956d17eb44dd367f9ae03460aa22e729a82c718)

amyeroberts reviewed (commented):
Thanks for all the work adding this model. 

Beautiful PR - really nice work. Just a few small comments to address but I think we're close to merge 🤗

sbucaille committed:
Apply suggestions from code review

Fixed small changes

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

Commit:
Moved ImagePointDescriptionOutput from modeling_outputs.py to modeling_superpoint.py

Commit:
Removed AutoModelForKeypointDetection and related stuff

Commit:
Fixed inconsistencies in image_processing_superpoint.py

Commit:
Moved infer_on_model logic simply in test_inference

Commit:
Fixed bugs, added labels to forward method with checks whether it is properly a None value, also added tests about this logic in test_modeling_superpoint.py

Commit:
Added tests to SuperPointImageProcessor to ensure that images are properly converted to grayscale

sbucaille commented:
Hey @amyeroberts ,
I addressed most of the suggestions, although I was not sure about one of them, let me know what you think. I'm looking forward this merge ! 😄 🤗

Commit:
Removed remaining mentions of MODEL_FOR_KEYPOINT_DETECTION_MAPPING

amyeroberts commented on src/transformers/models/superpoint/modeling_superpoint.py:
```diff
@@ -0,0 +1,510 @@
+    scores: Optional[torch.FloatTensor] = None
+    descriptors: Optional[torch.FloatTensor] = None
+    mask: Optional[torch.BoolTensor] = None
+    last_hidden_state: torch.FloatTensor = None
```
last_hidden_states should be the first returned value here

sbucaille replied:
Fixed in [this commit](https://github.com/huggingface/transformers/pull/28966/commits/c8a21216ac36870e0e87537a79de46593bb6f906)

amyeroberts reviewed (commented):
Looks great - thanks for iterating on this! 

Just a few small comments. The final thing to do before merging in is to transfer the checkpoints to under the Magic Leap org. I can't remember the outcome of this - did you   reach out to any of the authors to ask if they'd like to control the org on the hub (I don't believe it exists yet).

Commit:
Fixed from (w, h) to (h, w) as input for tests

sbucaille committed:
Apply suggestions from code review

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

Commit:
Removed unnecessary condition

Commit:
Moved last_hidden_state to be the first returned

Commit:
Moved last_hidden_state to be the first returned (bis)

Commit:
Moved last_hidden_state to be the first returned (ter)

Commit:
Switched image_width and image_height in tests to match recent changes

Commit:
Added config as first SuperPointConvBlock init argument

sbucaille commented on src/transformers/models/superpoint/modeling_superpoint.py:
```diff
@@ -0,0 +1,510 @@
+            scores=scores,
+            descriptors=descriptors,
+            mask=mask,
+            last_hidden_state=last_hidden_state,
```
Fixed in [this commit](https://github.com/huggingface/transformers/pull/28966/commits/c8a21216ac36870e0e87537a79de46593bb6f906)

sbucaille commented on tests/models/superpoint/test_modeling_superpoint.py:
```diff
@@ -0,0 +1,307 @@
+            with torch.no_grad():
+                outputs = model(**self._prepare_for_class(inputs_dict, model_class))
+
+            hidden_states = outputs.encoder_hidden_states if config.is_encoder_decoder else outputs.hidden_states
```
Fixed in [this commit](https://github.com/huggingface/transformers/pull/28966/commits/2d9444f1bda8b10fbbe049b18df511bf3d7d49b0)

sbucaille commented on tests/models/superpoint/test_modeling_superpoint.py:
```diff
@@ -0,0 +1,307 @@
+                    [
+                        conv_layer_size,
+                        self.model_tester.image_width // (2 ** (i + 1)),
+                        self.model_tester.image_height // (2 ** (i + 1)),
```
You are right, I always get confused in dimension orders, the input of the model is indeed (b, c, h, w). The tests are confusing, I fixed it all in [this commit](https://github.com/huggingface/transformers/pull/28966/commits/e0816c5f78e44618da2318672a6b38f1190998a6)

Commit:
Reordered README's after merge

Commit:
Added missing first config argument to SuperPointConvBlock instantiations

Commit:
Removed formatting error

sbucaille commented:
Hi @amyeroberts,
I addressed the last problems apart from the README one. I sent an email to the authors, waiting for their answer. But none of them still work at MagicLeap, in the case I don't have a positive answer from them and we can't have a repository owned by the company, what should we do ?

sbucaille commented:
Hi @amyeroberts ,
I managed to get the contact of a potential individual that would be in charge of this, I can add you to the mail thread if you give me an email address.
Regarding the weights, you can either copy them from the repo [I initially created ](https://huggingface.co/stevenbucaille/superpoint/tree/main) or run the `convert_superpoint_to_pytorch.py` script provided in the superpoint module

amyeroberts commented:
@sbucaille Great! You can get me on my HF email: amy@huggingface.co

amyeroberts commented:
@sbucaille I've copied across the checkpoints - so they can be updated in the PR now. 

Could you open a PR on https://huggingface.co/magic-leap-community/superpoint to fill in the model card?

sbucaille commented:
Hi @amyeroberts,  as you can probably see, I've added you to the mail thread and opened a pull request to the hub superpoint repo

Commit:
Added SuperPoint to README's de, pt-br, ru, te and vi

sbucaille commented:
Also, just rebased the branch and figured out main was not passing tests because of repo-consistency checks, I've added SuperPoint to the appropriate README's and will watch closely when main is updated to rebase once again with what it missing there

amyeroberts reviewed (approved):
Looks great - thanks for all your work on adding this model! 

Last things to do are rebase on main and trying to resolve the weird diff we're still seeing on README.fr

Commit:
Checked out README_fr.md

Commit:
Fixed README_fr.md

Commit:
Test fix README_fr.md

Commit:
Test fix README_fr.md

Commit:
Last make fix-copies !

amyeroberts commented on tests/models/superpoint/test_modeling_superpoint.py:
```diff
@@ -0,0 +1,307 @@
+class SuperPointModelIntegrationTest(unittest.TestCase):
+    @cached_property
+    def default_image_processor(self):
+        return AutoImageProcessor.from_pretrained("stevenbucaille/superpoint") if is_vision_available() else None
```
Checkpoint should be updated here

amyeroberts commented on docs/source/en/model_doc/superpoint.md:
```diff
@@ -0,0 +1,127 @@
+url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+image = Image.open(requests.get(url, stream=True).raw)
+
+processor = AutoImageProcessor.from_pretrained("stevenbucaille/superpoint")
+model = AutoModel.from_pretrained("stevenbucaille/superpoint")
```
Checkpoints should be updated here

amyeroberts commented on docs/source/en/model_doc/superpoint.md:
```diff
@@ -0,0 +1,127 @@
+
+images = [image_1, image_2]
+
+processor = AutoImageProcessor.from_pretrained("stevenbucaille/superpoint")
+model = AutoModel.from_pretrained("stevenbucaille/superpoint")
```
And here

amyeroberts commented on src/transformers/models/superpoint/configuration_superpoint.py:
```diff
@@ -0,0 +1,91 @@
+logger = logging.get_logger(__name__)
+
+SUPERPOINT_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "stevenbucaille/superpoint": "https://huggingface.co/stevenbucaille/superpoint/blob/main/config.json"
```
Checkpoint to be updated

amyeroberts commented on src/transformers/models/superpoint/configuration_superpoint.py:
```diff
@@ -0,0 +1,91 @@
+    This is the configuration class to store the configuration of a [`SuperPointModel`]. It is used to instantiate a
+    SuperPoint model according to the specified arguments, defining the model architecture. Instantiating a
+    configuration with the defaults will yield a similar configuration to that of the SuperPoint
+    [stevenbucaille/superpoint](https://huggingface.co/stevenbucaille/superpoint) architecture.
```
Same here

amyeroberts commented on tests/models/superpoint/test_modeling_superpoint.py:
```diff
@@ -0,0 +1,307 @@
+
+    @slow
+    def test_inference(self):
+        model = SuperPointModel.from_pretrained("stevenbucaille/superpoint").to(torch_device)
```
And here for the checkpoint

amyeroberts reviewed (approved):
Thanks for all the work adding this and digging into the funny README issue. 

Only thing left to do is update the checkpoints!

amyeroberts commented:
@sbucaille Awesome work adding this. If you can get the checkpoints updated for today we might be able to have this model make the next release 💪

Commit:
Updated checkpoint path

Commit:
Removed unused SuperPoint doc

sbucaille commented:
Hi @amyeroberts ,
Checkpoints are updated, I also fixed deleted remains of SuperPointForInterestPointDetection, test_pr_documentation test now passes. I guess we're good !

Commit:
Added missing image

sbucaille commented:
![image](https://github.com/huggingface/transformers/assets/24275548/146be8e0-ac24-460a-8c0a-4bfae6374ad4)

Is a screenshot sufficient ? When moving from the previous pull request to this one I forgot to add the additional COCO image.
Also, I'm not very familiar with the autodoc things but what causes this issue ? 
![image](https://github.com/huggingface/transformers/assets/24275548/296ef1f0-cbb4-48ec-bd63-667fab86efb7)

amyeroberts commented on src/transformers/models/superpoint/modeling_superpoint.py:
```diff
@@ -0,0 +1,516 @@
+        self.post_init()
+
+    @add_start_docstrings_to_model_forward(SUPERPOINT_INPUTS_DOCSTRING)
+    @add_code_sample_docstrings(
+        checkpoint=_CHECKPOINT_FOR_DOC,
+        output_type=ImagePointDescriptionOutput,
+        config_class=_CONFIG_FOR_DOC,
+        modality="vision",
+    )
```
@sbucaille The error is happening because there isn't an `expected_output` argument defined here

We should remove this as:
* This isn't really a "base model" [as per the code sample this creates](https://github.com/huggingface/transformers/blob/c852d4fba6a46f8b71c608041a10691b66f174e5/src/transformers/utils/doc.py#L541)
* There's already a code example provided in the docstring

```suggestion
```

amyeroberts commented:
> @amyeroberts Is a screenshot sufficient ?

Main thing I'd need to see is the end of the test run, which shows how many of the tests have passed / been skipped etc. (I. can't see the command being run, but make sure to set RUN_SLOW=1)

I've commented on the part causing the error for the doctests

sbucaille committed:
Update src/transformers/models/superpoint/modeling_superpoint.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

sbucaille commented:
@amyeroberts

```
RUN_SLOW=1 pytest tests/models/superpoint/
================================================================ test session starts =================================================================
platform linux -- Python 3.10.12, pytest-7.4.4, pluggy-1.4.0
rootdir: /home/steven.bucaille/transformers
configfile: pyproject.toml
plugins: dash-2.16.1, xdist-3.5.0, hypothesis-6.99.8, timeout-2.3.1
collected 93 items

tests/models/superpoint/test_image_processing_superpoint.py .s...........                                                                      [ 13%]
tests/models/superpoint/test_modeling_superpoint.py s........ssss....sssssssss.............s......s....sss.s.....s.......ss....ssss.           [100%]

================================================================== warnings summary ==================================================================
venv/lib/python3.10/site-packages/_pytest/config/__init__.py:1373
  /home/steven.bucaille/transformers/venv/lib/python3.10/site-packages/_pytest/config/__init__.py:1373: PytestConfigWarning: Unknown config option: doctest_glob

    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_fast_init_context_manager
  /home/steven.bucaille/transformers/tests/test_modeling_common.py:465: FutureWarning: `torch.testing.assert_allclose()` is deprecated since 1.12 and will be removed in a future release. Please use `torch.testing.assert_close()` instead. You can find detailed upgrade instructions in https://github.com/pytorch/pytorch/issues/61844.
    torch.testing.assert_allclose(init_instance.linear.bias, expected_bias, rtol=1e-3, atol=1e-4)

tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_fast_init_context_manager
  /home/steven.bucaille/transformers/tests/test_modeling_common.py:468: FutureWarning: `torch.testing.assert_allclose()` is deprecated since 1.12 and will be removed in a future release. Please use `torch.testing.assert_close()` instead. You can find detailed upgrade instructions in https://github.com/pytorch/pytorch/issues/61844.
    torch.testing.assert_allclose(

tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_attentions
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_hidden_state
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_simple
  /home/steven.bucaille/transformers/src/transformers/modeling_utils.py:4225: FutureWarning: `_is_quantized_training_enabled` is going to be deprecated in transformers 4.39.0. Please use `model.hf_quantizer.is_trainable` instead
    warnings.warn(

tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_attentions
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_hidden_state
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_simple
  /home/steven.bucaille/transformers/src/transformers/models/superpoint/modeling_superpoint.py:475: TracerWarning: Iterating over a tensor might cause the trace to be incorrect. Passing a tensor of different shape won't change the number of iterations executed (and might lead to errors or silently give incorrect results).
    list_keypoints_scores = [

tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_attentions
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_hidden_state
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_simple
  /home/steven.bucaille/transformers/src/transformers/models/superpoint/modeling_superpoint.py:249: TracerWarning: Iterating over a tensor might cause the trace to be incorrect. Passing a tensor of different shape won't change the number of iterations executed (and might lead to errors or silently give incorrect results).
    scores = scores[0][tuple(keypoints.t())]

tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_attentions
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_hidden_state
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_simple
  /home/steven.bucaille/transformers/src/transformers/models/superpoint/modeling_superpoint.py:249: TracerWarning: Using len to get tensor shape might cause the trace to be incorrect. Recommended usage would be tensor.shape[0]. Passing a tensor of different shape might lead to errors or silently give incorrect results.
    scores = scores[0][tuple(keypoints.t())]

tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_attentions
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_hidden_state
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_simple
  /home/steven.bucaille/transformers/src/transformers/models/superpoint/modeling_superpoint.py:484: TracerWarning: Iterating over a tensor might cause the trace to be incorrect. Passing a tensor of different shape won't change the number of iterations executed (and might lead to errors or silently give incorrect results).
    for last_hidden_state, keypoints in zip(last_hidden_state, list_keypoints)

tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_attentions
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_hidden_state
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_simple
  /home/steven.bucaille/transformers/src/transformers/models/superpoint/modeling_superpoint.py:312: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
    divisor = torch.tensor([[(width * scale - scale / 2 - 0.5), (height * scale - scale / 2 - 0.5)]])

tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_attentions
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_hidden_state
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_simple
  /home/steven.bucaille/transformers/src/transformers/models/superpoint/modeling_superpoint.py:312: TracerWarning: Converting a tensor to a Python float might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
    divisor = torch.tensor([[(width * scale - scale / 2 - 0.5), (height * scale - scale / 2 - 0.5)]])

tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_attentions
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_output_hidden_state
tests/models/superpoint/test_modeling_superpoint.py::SuperPointModelTest::test_torchscript_simple
  /home/steven.bucaille/transformers/src/transformers/models/superpoint/modeling_superpoint.py:487: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
    maximum_num_keypoints = max(keypoints.shape[0] for keypoints in list_keypoints)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
==================================================== 65 passed, 28 skipped, 27 warnings in 4.19s =====================================================
```

I applied your changes on the doc

Commit:
Removed unnecessary import

ydshieh commented:
(thank you for requesting running slow tests @amyeroberts  and thank you for running the slow tests @sbucaille ❤️ !)

sbucaille commented:
@amyeroberts I'm afraid the error is back :thinking:

amyeroberts commented:
@sbucaille I was wrong about the fix then! I'll check out your branch and see if I can figure out what's happening

ydshieh commented:
> @sbucaille I was wrong about the fix then! I'll check out your branch and see if I can figure out what's happening

don't hesitate to ping me if you have limited bandwidth @amyeroberts

amyeroberts commented on src/transformers/models/superpoint/modeling_superpoint.py:
```diff
@@ -0,0 +1,509 @@
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, ImagePointDescriptionOutput]:
+        """
+        Returns:
+
```
ok - I actually think this is the offending line

```suggestion
```

sbucaille committed:
Update src/transformers/models/superpoint/modeling_superpoint.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

sbucaille commented:
@amyeroberts should I revert the previous change ?

amyeroberts commented:
@sbucaille For the code example? No, that also wasn't correct (I just missed it in my reviews 🙈 )

amyeroberts commented:
OK - new issue. This one I think is because there isn't a model entry added in `docs/source/en/_toctree.yml`

Commit:
Added SuperPoint to _toctree.yml

sbucaille commented:
@amyeroberts Added it to _toctree, I believe this belongs to the vision section ?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_28966). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

amyeroberts commented:
@sbucaille Yep! And all passing now! Thanks for all your work on this - great to see this added to the library 🤗

sbucaille commented:
Yaaay finally ! :joy: Thanks for the opportunity and all the feedback provided during this long lasting PR, I learned a lot of things !
Time to implement SuperGlue now :sunglasses: :hugs:
